### DISCUSSION TITLE
Create a standard for representing doorbells

### DISCUSSION THREAD

**Discussion by bdraco** [home-assistant/architecture#869] [Entity Models] [answered]
Currently most doorbells have a device class of `BinarySensorDeviceClass.OCCUPANCY`

This is a bit confusing as we don't know how to separate doorbells from non-doorbells. This leads to some confusion.

The goal here is to resolve how to represent a doorbell in a way that can:

- Known to be a doorbell both by the user and programmatically so it can be consumed by other integrations
- Easily track and view history of the doorbell presses

https://github.com/home-assistant/core/pull/87545#discussion_r1097574627

Potential implementation for discussion https://github.com/home-assistant/core/pull/87559

**Comment by frenck** [comment]
I agree a separate device class should be a good idea. However...

Doorbells are momentary button pushes, thus stateless. Their bare basics are not to be a `binary_sensor`, but an event.

For that reason, I don't directly see why this should be added, as it may lead to incorrect usage of binary sensors (as IMHO is already handled incorrectly in the linked PR).

../Frenck

**Reply by bdraco** [reply]
We currently have most of them implemented as binary sensors in the codebase

Added a sample PR https://github.com/home-assistant/core/pull/87559

**Reply by frenck** [reply]
Right, but why would we allow doorbell pushes to be binary sensors, but button pushes on e.g., Philips Hue remotes aren't?

That makes no sense to me. IMHO, it remains that we shouldn't add this device class, as it provides an incentive to use practices that are in conflict with our general design standards.

../Frenck

**Reply by bdraco** [reply]
They could all be migrated to device triggers, but that would make them much harder to use as there is no way to categorize a device trigger since they lack any type of system to standardize what a device trigger does.  Ie we have no trigger classes etc.

**Reply by AngellusMortis** [reply]
The issue with them as events is that there is no way to track them. Users want to know when the last time the doorbell was pressed Events/device trackers have no way for a user to add them to a dashboard. The logbook is not really an acceptable alternative because it is a scrolling list. 

This could also be solved with a "last rang" datetime sensor, but as @bdraco said, the binary sensor has been the convention for a while, and I think that is what the HomeKit integration currently is looking for. Ideally though, events and device triggers should be improved to make it easier for users to discover them and add them to dashboards. And yeah, there needs to be a way to categorize device triggers so you can identify doorbells specifically (again, important for something like HomeKit).

**Reply by frenck** [reply]
That is a generic concern, not related to doorbells specifically. This is already an existing problem. I'm not sure what that has to do with this device class.

Binary sensors have a binary state, momentary buttons do not have such a state, and adding a device class for a doorbell is therefore not right.

**Reply by allenporter** [reply]
The fact that we have some doorbells only using events (e.g. nest) and others still allowed as binary sensors definitely results in lots of end user confusion as users have mismatched expectations and like the practical usability of binary sensors.  Then users make template binary sensors to recreate the behavior. Just highlighting the problem here that we should resolve this misalignment in some way.



Is the thinking that we'd want to have easier ways to make virtual binary sensors based on events more generally for help in automations? (My impression is some core members had a discussion on this a year or more ago) Or have new ways to display events in the history/log book?

**Reply by frenck** [reply]
Oh, I agree with the general concerns, @allenporter. I personally don't like to use events either. But that is a different discussion to which in the past was agreed on adding device automation. I personally would applaud just entities. Which would be so much easier to use in general.

But, that is circling around a different issue and discussion; the proposal for this architectural issue was: To add a doorbell device class.

The fact remains, that binary sensor are not made/designed for the use case that this device class suggested in this architectural issue.

**Reply by thecode** [reply]
My first thought was yeah, why not add a new device class, than I realized we don't allow binary sensors for momentary events, I think the main problem is that some integrations succeed to add binary sensors for momentary events which created confusion.
If we add such device class, what preventing any integration which wants to create a sensor for a momentary event create it as a doorbell?

There is however one thing I would like to note, which I don't know if the linked PR integration is working like this, some doorbells will let the user do a short press (momentary) but the press status will stay for a while, but this might be fitting more to a "call" sensor.

**Reply by bdraco** [reply]
> There is however one thing I would like to note, which I don't know if the linked PR integration is working like this, some doorbells will let the user do a short press (momentary) but the press status will stay for a while, but this might be fitting more to a "call" sensor.

This one is also a little bit muddy because you can hold down some doorbells and they stay in a pressed state.  

We need a single way to represent a doorbell as the current bifurcated state of things leads to a poor dev and user experience since there is no clear direction on how to proceed.  

Right now there are effectively two options:

- binary sensor with a confusing device class

- device trigger/event path which has less functionality than the binary sensor so its not a full replacement 

This leaves the devs implementing these stuck in the middle with only bad choices since the device class is confusing, and with events they cannot easily track history or link to voice assistants without some manual work which lands them issues in their issue queue with no good answer.

**Reply by frenck** [reply]
> Right now there are effectively two options:

That, is thus, something I don't agree on. As we officially don't allow momentary buttons to be represented as binary sensors. 🤷 

But neither of the above is related to the things this architectural proposal raises. Nor does adding a device class solve the general other concerns raised.

../Frenck

**Reply by bdraco** [reply]
Let's set aside the original proposal for the purposes of continuing this discussion as I don't think it particularly matters if the outcome is a new device class.  The goal here is to resolve how to represent a doorbell in a way that can:

- Known to be a doorbell both by the user and programmatically so it can be consumed by other integrations
- Easily track and view history of the doorbell presses

**Reply by frenck** [reply]
The purpose of this discussion was to add a device class and it has been massively derailed.

See also the linked PR in the proposal: <https://github.com/home-assistant/core/pull/87559>

**Reply by bdraco** [reply]
Short of closing this out and starting another discussion, I'm thinking it's best to reframe the discussion as it's uncovering larger issues (that's a good thing) that we need to be resolved first and I didn't want to loose context about how we got here.

> See also the linked PR in the proposal: [home-assistant/core#87559](https://github.com/home-assistant/core/pull/87559)

If that turns out not to be the path forward, thats fine with me, as the goal of the PR was to provide one possible solution for this issue.

**Reply by elupus** [reply]
Old school doorbells are not one event, they are two. Pressed and released. Ding and Dong style. Could obviously be mapped to two events. But i do like the simplicity of binary sensor.

**Reply by thecode** [reply]
@elupus how would that be different than any button which support press and release events? For example https://www.home-assistant.io/integrations/shelly/#possible-values-for-click_type we have a `btn_up` and `btn_down` events since this is a momentary button (I would prefer it to be a binary sensor) I don't want to drift from the main subject of propousing device class, but ideally we can think of a solution that will fit all type of momentary devices. For example create some type of sensor that would be used for these cases.

**Reply by elupus** [reply]
Not against that at all. I prefer an actual entity too. It does fit a binary sensor, harder with something that is just a single event

**Reply by Djelibeybi** [reply]
Why isn't this an Entity Category for the Button entity, i.e. along with Restart and Upgrade we have Doorbell for Buttons?

**Reply by bagobones** [reply]
Ways to look at it.

A traditional door bell:
Is a momentary button with a down and up event.

A zwave doorbell like the aeotec doorbell
presents it self as an alarm / siren object on and off.

Homekit /  ONVIF / Video / Smart Doorbells:
Are cameras with motion/person sensors and button/occupancy sensors
IE someone is near the door, vs someone is AT the door / has pushed the button

A major missing item in the current GUI homekit setup is the ability to map the motion and button/occupancy sensors.. You can do it in YAML but not with the GUI setup.

With ONVIF profile T now we may get more local two way capable door bells, these devices again are "security cameras" with buttons and motion features. At least that is how ONVIF treats them but no official standard on how to handle "the button"

**Reply by starkillerOG** [reply]
For the reolink doorbell we already had a discussion about the visitor sensor beeing a binary_sensor vs events:
https://github.com/home-assistant/core/pull/85654#discussion_r1082655476

There we agreeded that it could be a binary_sensor because reolink implements it that way:
When you press the doorbell, you get a visitor=TRUE notification, about 15 seconds later you get a second visitor=FALSE notification.
You can also pol the doorbell button status with the API and it will return visitor=TRUE/FALSE (within the 15 seconds after a press it will remain TRUE).

So clearly inside the reolink camera firmware the button is implemented as a binary TRUE/FALSE value that will keep TRUE for a while.

**Reply by Kane610** [reply]
The button on Axis doorstations sends boolean states and it stays true while the button is kept pressed.

**Reply by Rudd-O** [reply]
I think we can and we should accommodate two _kinds_ of doorbells:

* traditional momentary type switches -> button press with specific device class doorbell
* newfangled bidirectional or modern doorbells -> binary sensor with specific device class doorbell or stateful doorbell, that can stay on until acked or somehow else acted upon

**Reply by bagobones** [reply]
I would suggest:

- doorbell (single sensor)
- video_doorbell (camera with multiple sensors, optional two way audio)

**Reply by frenck** [reply]
> So clearly inside the reolink camera firmware the button is implemented as a binary TRUE/FALSE value that will keep TRUE for a while.

The problem is the end-user experience that is just bad across HA regarding what is (physically to the user) a momentary switch. No matter how the API or device is implemented it, the results are a variety different methods across how things work in HA, while our goal is to make everything work the same.

That is not right and needs to be fixed. The physical button is momentary, it should be represented as such in HA as well. Which right now is an event by our standards, not a binary sensor.

You can't explain to the average user why one clicky button is a binary sensor and why another is an event. That is just weird and wrong.

../Frenck


PS: I've been working on an extensive proposal/solution the past week to deal with all the above, which might be able to solve it for all cases (not just doorbells). Once I got it all written up, I'll write up a discussion in this repo.

**Reply by Rudd-O** [reply]
> PS: I've been working on an extensive proposal/solution the past week to deal with all the above, which might be able to solve it for all cases (not just doorbells). Once I got it all written up, I'll write up a discussion in this repo.

Super stoked to read it as soon as it is published.

**Reply by frenck** [reply, marked as answer]
Architectural counter-proposal: <https://github.com/home-assistant/architecture/discussions/929>
